### PR TITLE
rubicon project adapter

### PR DESCRIPTION
**Pull request by mhc-gh**
Initial submission of Rubicon Project adapter

### CODE REVIEW

**Comment by cramforce** [comment]
LGTM, but please let me know whether local testing ended up working

**Comment by mhc-gh** [comment]
I added some bug fixes for the FastLane part of the adapter so that's no working as expected and supporting documentation is updated to reflect the slight configuration change.

File: ads/rubicon.js [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+  let gptran = false;
+  function gptrun() {
+    if (gptran) {
+      return;
+    }
+    gptran = true;
+
+    let ASTargeting = rubicontag.getSlot('c').getAdServerTargeting();
+    const ptrn = /rpfl_\d+/i;
+    for (i = 0; i < ASTargeting.length; i++) {
```
Comment (cramforce) [inline ads/rubicon.js]: let i

File: ads/rubicon.js [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    });
+  });
+}
+
+/**
+ * @param {!Window} global
+ * @param {!Object} data
+ */
+function smartTag(global, data) {
+  const pageURL = context.location.href.replace('cdn.ampproject.org/c/s/','');
```
Comment (cramforce) [inline ads/rubicon.js]: This can contain other prefixes.

Take a look at `getSourceUrl` from url.js

**Comment by cramforce** [comment]
A few more comments. Please squash commits.

**Comment by mhc-gh** [comment]
Can you give this the once over again and then once confirmed okay i'll look to try squashing the commits.

**Comment by cramforce** [comment]
LGTM

**Comment by mhc-gh** [comment]
I am following this but it's not squashing...
http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html

**Comment by cramforce** [comment]
Basically what you want to do is:

- `git rebase -i`
- Enter `s` instead of `pick` on all but one line
- Save

**Comment by mhc-gh** [comment]
Yeah, that's what I did.  It seemed to work and then it added a new commit (ecdc18c) and a merge (0dfcc97) and left the old ones in the pull request also.

**Comment by Meggin** [comment]
Maybe missing this step?

git push —force

For what it's worth, took me awhile to get use to squashing, and now I love
it. Clean.

**Comment by cramforce** [comment]
There is also `git commit --amend` for the lazy :)

**Comment by mhc-gh** [comment]
Bah! Still not working.  No idea what's going wrong.  Is it possible to just merge in this pull request as is?

**Comment by mhc-gh** [comment]
This is what I see:

git log --pretty=format:"%h - %an, %ar : %s" -5
5aa6415 - Martin Hill, 50 minutes ago : Merge branch 'master' of https://github.com/mhc-gh/amphtml merge to rebase commits for pull request
ecdc18c - Martin Hill, 23 hours ago : adding rubicon adapter
33cd993 - Martin Hill, 63 minutes ago : adapter updates as per comments
574d092 - Martin Hill, 2 hours ago : rubicon adapter bug fixes
27518a3 - Martin Hill, 23 hours ago : adding rubicon adapter

git rebase -i HEAD~5
pick 028bfc3 Use location.replace for click interception and only when iframed
pick 72a6a67 Allow using preloading over prefetching.
pick a29982a update links to new ampbyexample.com domain
pick ecdc18c adding rubicon adapter
pick 27518a3 adding rubicon adapter
pick 574d092 rubicon adapter bug fixes
pick 33cd993 adapter updates as per comments

# Rebase 9dccb37..5aa6415 onto 9dccb37 (7 command(s))
#
# Commands:
# p, pick = use commit
# r, reword = use commit, but edit the commit message
# e, edit = use commit, but stop for amending
# s, squash = use commit, but meld into previous commit
# f, fixup = like "squash", but discard this commit's log message
# x, exec = run command (the rest of the line) using shell
#
# These lines can be re-ordered; they are executed from top to bottom.
#
# If you remove a line here THAT COMMIT WILL BE LOST.
#
# However, if you remove everything, the rebase will be aborted.
#
# Note that empty commits are commented out

So now rebasing seems to mis-match the last commits and wants me to bring in others.  I'm not sure what I should do to resolve it?

**Comment by cramforce** [comment]
Is your branch tracking master. Make sure to rebase first and then do the interactive rebase when the actual delta with master is 0.

**Comment by mhc-gh** [comment]
Okay, looks like I made it work.

**Comment by mhc-gh** [comment]
Not sure why the travis-ci build has failed but looks like it's unrelated to my actual pull request.  Can you verify and confirm so we can proceed with the PR?

**Comment by mhc-gh** [comment]
I closed and reopened the PR in order to have the build re-run without checking in anything new.  It's now passed successfully.

**Comment by cramforce** [comment]
LGTM

**Pull request merged**